src/services/claim.ts

```diff
@@ -0,0 +1,183 @@
+}
+
+export async function getPendingClaims(
+  network: NetworkId,
+  provider: Web3Provider,
+  account: string
+): Promise<{ claims: Claim[]; reports: Report }> {
+  const snapshot = await getSnapshot(network);
+
+  const claimStatus = await getClaimStatus(
+    network,
+    provider,
+    Object.keys(snapshot).length,
+    account
+  );
+
+  const pendingWeeks = claimStatus
+    .map((status, i) => [i + 1, status])
+    .filter(([, status]) => !status)
+    .map(([i]) => i) as number[];
+
+  const pendingWeeksReports = await getReports(snapshot, pendingWeeks);
+
+  return {
+    claims: Object.entries(pendingWeeksReports)
+      .filter((report: Report) => report[1][account])
+      .map((report: Report) => {
+        return {
+          id: report[0],
+          amount: report[1][account],
+          amountDenorm: parseUnits(report[1][account], 18)
+        };
+      }),
+    reports: pendingWeeksReports
+  };
+}
+
+type CurrentRewardsEstimateResponse = {
+  success: boolean;
+  result: {
+    current_timestamp: string;
+    'liquidity-providers': Array<{
+      snapshot_timestamp: string;
+      address: string;
+      current_estimate: string;
+      velocity: string;
+      week: number;
+    }>;
+  };
+};
+
+export type CurrentRewardsEstimate = {
+  rewards: string;
+  velocity: string;
+  timestamp: string;
+} | null;
+
+export async function getCurrentRewardsEstimate(
+  account: string
+): Promise<CurrentRewardsEstimate> {
+  try {
+    const response = await axios.get<CurrentRewardsEstimateResponse>(
+      `https://api.balancer.finance/liquidity-mining/v1/liquidity-provider/${account}`
+    );
+    if (response.data.success) {
+      const liquidityProviders = response.data.result['liquidity-providers'];
+      const rewards = liquidityProviders
+        .reduce(
+          (total, { current_estimate }) => total.plus(current_estimate),
+          bnum(0)
+        )
+        .toString();
+      const velocity =
+        liquidityProviders
+          .find(liquidityProvider => Number(liquidityProvider.velocity) > 0)
+          ?.velocity.toString() ?? '0';
+
+      if (Array.isArray(liquidityProviders)) {
+        return {
+          rewards,
+          velocity,
+          timestamp: response.data.result.current_timestamp
+        };
+      }
+    }
+  } catch (e) {
+    console.log('[Claim] Current Rewards Estimate Error', e);
+  }
+  return null;
+}
+
+export async function claimRewards(
+  network: NetworkId,
+  provider: Web3Provider,
+  account: string,
+  pendingClaims: Claim[],
+  reports: Report
+): Promise<TransactionResponse> {
+  try {
+    const claims = pendingClaims.map(week => {
+      const claimBalance = week.amount;
+      const merkleTree = loadTree(reports[week.id]);
+
+      const proof = merkleTree.getHexProof(
+        soliditySha3(account, toWei(claimBalance))
+      );
+      return [parseInt(week.id), toWei(claimBalance), proof];
+    });
```

Comment:
This function is pretty slow and can even cause the browser to freeze if you have a lot to claim... this is a pretty heavy computation... I do have some experience with web workers which might be the solution in this case. (it will definitely help the UI/UX and not "freeze" the browser)

Reply:
I think it would be better to implement it under a web worker then right?

Reply:
I think it would be better to instil a better trust and confidence in the UI if we don't have issues with lag or freezes from the get go

Reply:
If we have time to try a web worker for this, then lets at least experiment with it. Or we could try a serverless function somewhere, but that complicates decentralising.